// @tamagui/select/dist/jsx/context.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContextScope } from "@tamagui/create-context";
import { SELECT_NAME } from "./constants";
function _array_like_to_array(arr, len) {
  (len == null || len > arr.length) && (len = arr.length);
  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];
  return arr2;
}
function _array_with_holes(arr) {
  if (Array.isArray(arr)) return arr;
}
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _iterable_to_array_limit(arr, i) {
  var _i = arr == null ? null : typeof Symbol < "u" && arr[Symbol.iterator] || arr["@@iterator"];
  if (_i != null) {
    var _arr = [], _n = !0, _d = !1, _s, _e;
    try {
      for (_i = _i.call(arr); !(_n = (_s = _i.next()).done) && (_arr.push(_s.value), !(i && _arr.length === i)); _n = !0)
        ;
    } catch (err) {
      _d = !0, _e = err;
    } finally {
      try {
        !_n && _i.return != null && _i.return();
      } finally {
        if (_d) throw _e;
      }
    }
    return _arr;
  }
}
function _non_iterable_rest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function _sliced_to_array(arr, i) {
  return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _unsupported_iterable_to_array(o, minLen) {
  if (o) {
    if (typeof o == "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor && (n = o.constructor.name), n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
  }
}
var _createContextScope = _sliced_to_array(createContextScope(SELECT_NAME), 2), createSelectContext = _createContextScope[0], createSelectScope = _createContextScope[1], _createSelectContext = _sliced_to_array(createSelectContext(SELECT_NAME), 2), SelectProvider = _createSelectContext[0], useSelectContext = _createSelectContext[1], _createContextScope1 = _sliced_to_array(createContextScope(SELECT_NAME), 2), createSelectItemParentContext = _createContextScope1[0], createSelectItemParentScope = _createContextScope1[1], _createSelectContext1 = _sliced_to_array(createSelectContext(SELECT_NAME), 2), SelectItemParentProvider = _createSelectContext1[0], useSelectItemParentContext = _createSelectContext1[1], ForwardSelectContext = function(param) {
  var __scopeSelect = param.__scopeSelect, context = param.context, itemContext = param.itemContext, children = param.children;
  return /* @__PURE__ */ _jsx(SelectProvider, _object_spread_props(_object_spread({
    isInSheet: !0,
    scope: __scopeSelect
  }, context), {
    children: /* @__PURE__ */ _jsx(SelectItemParentProvider, _object_spread_props(_object_spread({
      scope: __scopeSelect
    }, itemContext), {
      children
    }))
  }));
};
export {
  ForwardSelectContext,
  SelectItemParentProvider,
  SelectProvider,
  createSelectContext,
  createSelectItemParentContext,
  createSelectItemParentScope,
  createSelectScope,
  useSelectContext,
  useSelectItemParentContext
};
