// react-native-web-lite/dist/esm/ActivityIndicator/index.native.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { StyleSheet } from "react-native-web-internals";
import View from "../View/index";
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function _object_without_properties(source, excluded) {
  if (source == null) return {};
  var target = _object_without_properties_loose(source, excluded), key, i;
  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
    for (i = 0; i < sourceSymbolKeys.length; i++)
      key = sourceSymbolKeys[i], !(excluded.indexOf(key) >= 0) && Object.prototype.propertyIsEnumerable.call(source, key) && (target[key] = source[key]);
  }
  return target;
}
function _object_without_properties_loose(source, excluded) {
  if (source == null) return {};
  var target = {}, sourceKeys = Object.keys(source), key, i;
  for (i = 0; i < sourceKeys.length; i++)
    key = sourceKeys[i], !(excluded.indexOf(key) >= 0) && (target[key] = source[key]);
  return target;
}
var createSvgCircle = function(style) {
  return /* @__PURE__ */ _jsx("circle", {
    cx: "16",
    cy: "16",
    fill: "none",
    r: "14",
    strokeWidth: "4",
    style
  });
}, ActivityIndicator = /* @__PURE__ */ React.forwardRef(function(props, forwardedRef) {
  var _props_animating = props.animating, animating = _props_animating === void 0 ? !0 : _props_animating, _props_color = props.color, color = _props_color === void 0 ? "#1976D2" : _props_color, _props_hidesWhenStopped = props.hidesWhenStopped, hidesWhenStopped = _props_hidesWhenStopped === void 0 ? !0 : _props_hidesWhenStopped, _props_size = props.size, size = _props_size === void 0 ? "small" : _props_size, style = props.style, other = _object_without_properties(props, [
    "animating",
    "color",
    "hidesWhenStopped",
    "size",
    "style"
  ]), svg = /* @__PURE__ */ _jsxs("svg", {
    height: "100%",
    viewBox: "0 0 32 32",
    width: "100%",
    children: [
      createSvgCircle({
        stroke: color,
        opacity: 0.2
      }),
      createSvgCircle({
        stroke: color,
        strokeDasharray: 80,
        strokeDashoffset: 60
      })
    ]
  });
  return /* @__PURE__ */ _jsx(View, _object_spread_props(_object_spread({}, other), {
    accessibilityRole: "progressbar",
    accessibilityValueMax: 1,
    accessibilityValueMin: 0,
    ref: forwardedRef,
    style: [
      styles.container,
      style
    ],
    children: /* @__PURE__ */ _jsx(View, {
      children: svg,
      style: [
        typeof size == "number" ? {
          height: size,
          width: size
        } : indicatorSizes[size],
        styles.animation,
        !animating && styles.animationPause,
        !animating && hidesWhenStopped && styles.hidesWhenStopped
      ]
    })
  }));
});
ActivityIndicator.displayName = "ActivityIndicator";
var styles = StyleSheet.create({
  container: {
    alignItems: "center",
    justifyContent: "center"
  },
  hidesWhenStopped: {
    visibility: "hidden"
  },
  animation: {
    animationDuration: "0.75s",
    animationKeyframes: [
      {
        "0%": {
          transform: [
            {
              rotate: "0deg"
            }
          ]
        },
        "100%": {
          transform: [
            {
              rotate: "360deg"
            }
          ]
        }
      }
    ],
    animationTimingFunction: "linear",
    animationIterationCount: "infinite"
  },
  animationPause: {
    animationPlayState: "paused"
  }
}), indicatorSizes = StyleSheet.create({
  small: {
    width: 20,
    height: 20
  },
  large: {
    width: 36,
    height: 36
  }
}), ActivityIndicator_default = ActivityIndicator;
export {
  ActivityIndicator_default as default
};
