// react-native-web-lite/dist/esm/SafeAreaView/index.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from "react";
import { StyleSheet, canUseDOM } from "react-native-web-internals";
import View from "../View/index";
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function _object_without_properties(source, excluded) {
  if (source == null) return {};
  var target = _object_without_properties_loose(source, excluded), key, i;
  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
    for (i = 0; i < sourceSymbolKeys.length; i++)
      key = sourceSymbolKeys[i], !(excluded.indexOf(key) >= 0) && Object.prototype.propertyIsEnumerable.call(source, key) && (target[key] = source[key]);
  }
  return target;
}
function _object_without_properties_loose(source, excluded) {
  if (source == null) return {};
  var target = {}, sourceKeys = Object.keys(source), key, i;
  for (i = 0; i < sourceKeys.length; i++)
    key = sourceKeys[i], !(excluded.indexOf(key) >= 0) && (target[key] = source[key]);
  return target;
}
var cssFunction = function() {
  return canUseDOM && window.CSS && window.CSS.supports && window.CSS.supports("top: constant(safe-area-inset-top)") ? "constant" : "env";
}(), SafeAreaView = /* @__PURE__ */ React.forwardRef(function(props, ref) {
  var style = props.style, rest = _object_without_properties(props, [
    "style"
  ]);
  return /* @__PURE__ */ _jsx(View, _object_spread_props(_object_spread({}, rest), {
    ref,
    style: StyleSheet.compose(styles.root, style)
  }));
});
SafeAreaView.displayName = "SafeAreaView";
var styles = StyleSheet.create({
  root: {
    paddingTop: "".concat(cssFunction, "(safe-area-inset-top)"),
    paddingRight: "".concat(cssFunction, "(safe-area-inset-right)"),
    paddingBottom: "".concat(cssFunction, "(safe-area-inset-bottom)"),
    paddingLeft: "".concat(cssFunction, "(safe-area-inset-left)")
  }
}), SafeAreaView_default = SafeAreaView;
export {
  SafeAreaView_default as default
};
