// react-native-web-lite/dist/esm/DeviceInfo/index.native.js
import { canUseDOM } from "react-native-web-internals";
import Dimensions from "../Dimensions/index";
var DeviceInfo = {
  Dimensions: {
    get windowPhysicalPixels() {
      var _Dimensions_get = Dimensions.get("window"), width = _Dimensions_get.width, height = _Dimensions_get.height, fontScale = _Dimensions_get.fontScale, scale = _Dimensions_get.scale;
      return {
        width: width * scale,
        height: height * scale,
        scale,
        fontScale
      };
    },
    get screenPhysicalPixels() {
      var _Dimensions_get1 = Dimensions.get("screen"), width1 = _Dimensions_get1.width, height1 = _Dimensions_get1.height, fontScale1 = _Dimensions_get1.fontScale, scale1 = _Dimensions_get1.scale;
      return {
        width: width1 * scale1,
        height: height1 * scale1,
        scale: scale1,
        fontScale: fontScale1
      };
    }
  },
  get locale() {
    if (canUseDOM)
      return navigator.languages ? navigator.languages[0] : navigator.language;
  },
  get totalMemory() {
    return canUseDOM ? navigator.deviceMemory : void 0;
  },
  get userAgent() {
    return canUseDOM ? navigator.userAgent : "";
  }
}, DeviceInfo_default = DeviceInfo;
export {
  DeviceInfo_default as default
};
