// @tamagui/create-theme/dist/esm/applyMask.native.js
import { createTheme } from "./createTheme";
import { getThemeInfo, setThemeInfo } from "./themeInfo";
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function applyMask(theme, mask) {
  var options = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {}, parentName = arguments.length > 3 ? arguments[3] : void 0, nextName = arguments.length > 4 ? arguments[4] : void 0, info = getThemeInfo(theme, parentName);
  if (!info)
    throw new Error(process.env.NODE_ENV !== "production" ? "No info found for theme, you must pass the theme created by createThemeFromPalette directly to extendTheme" : "\u274C Err2");
  var next = applyMaskStateless(info, mask, options, parentName);
  return setThemeInfo(next.theme, {
    definition: next.definition,
    palette: info.palette,
    name: nextName
  }), next.theme;
}
function applyMaskStateless(info, mask) {
  var options = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {}, parentName = arguments.length > 3 ? arguments[3] : void 0, _info_options, skip = _object_spread({}, options.skip);
  if (!((_info_options = info.options) === null || _info_options === void 0) && _info_options.nonInheritedValues)
    for (var key in info.options.nonInheritedValues)
      skip[key] = 1;
  var maskOptions = _object_spread_props(_object_spread({
    parentName,
    palette: info.palette
  }, options), {
    skip
  }), template = mask.mask(info.definition, maskOptions), theme = createTheme(info.palette, template);
  return _object_spread_props(_object_spread({}, info), {
    cache: /* @__PURE__ */ new Map(),
    definition: template,
    theme
  });
}
export {
  applyMask,
  applyMaskStateless
};
