// react-native-web-lite/dist/cjs/Image/index.native.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf, __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: !0 }) : target,
  mod
)), __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var Image_exports = {};
__export(Image_exports, {
  default: () => Image_default
});
module.exports = __toCommonJS(Image_exports);
var import_jsx_runtime = require("react/jsx-runtime"), React = __toESM(require("react")), import_react_native_web_internals = require("react-native-web-internals"), import_react_native_web_internals2 = require("react-native-web-internals"), import_createElement = __toESM(require("../createElement/index")), import_PixelRatio = __toESM(require("../PixelRatio/index")), import_View = __toESM(require("../View/index"));
function _array_like_to_array(arr, len) {
  (len == null || len > arr.length) && (len = arr.length);
  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];
  return arr2;
}
function _array_with_holes(arr) {
  if (Array.isArray(arr)) return arr;
}
function _array_without_holes(arr) {
  if (Array.isArray(arr)) return _array_like_to_array(arr);
}
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _iterable_to_array(iter) {
  if (typeof Symbol < "u" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter);
}
function _iterable_to_array_limit(arr, i) {
  var _i = arr == null ? null : typeof Symbol < "u" && arr[Symbol.iterator] || arr["@@iterator"];
  if (_i != null) {
    var _arr = [], _n = !0, _d = !1, _s, _e;
    try {
      for (_i = _i.call(arr); !(_n = (_s = _i.next()).done) && (_arr.push(_s.value), !(i && _arr.length === i)); _n = !0)
        ;
    } catch (err) {
      _d = !0, _e = err;
    } finally {
      try {
        !_n && _i.return != null && _i.return();
      } finally {
        if (_d) throw _e;
      }
    }
    return _arr;
  }
}
function _non_iterable_rest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _non_iterable_spread() {
  throw new TypeError("Invalid attempt to spread non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function _object_without_properties(source, excluded) {
  if (source == null) return {};
  var target = _object_without_properties_loose(source, excluded), key, i;
  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
    for (i = 0; i < sourceSymbolKeys.length; i++)
      key = sourceSymbolKeys[i], !(excluded.indexOf(key) >= 0) && Object.prototype.propertyIsEnumerable.call(source, key) && (target[key] = source[key]);
  }
  return target;
}
function _object_without_properties_loose(source, excluded) {
  if (source == null) return {};
  var target = {}, sourceKeys = Object.keys(source), key, i;
  for (i = 0; i < sourceKeys.length; i++)
    key = sourceKeys[i], !(excluded.indexOf(key) >= 0) && (target[key] = source[key]);
  return target;
}
function _sliced_to_array(arr, i) {
  return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _to_consumable_array(arr) {
  return _array_without_holes(arr) || _iterable_to_array(arr) || _unsupported_iterable_to_array(arr) || _non_iterable_spread();
}
function _unsupported_iterable_to_array(o, minLen) {
  if (o) {
    if (typeof o == "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor && (n = o.constructor.name), n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
  }
}
var ERRORED = "ERRORED", LOADED = "LOADED", LOADING = "LOADING", IDLE = "IDLE", _filterId = 0, svgDataUriPattern = /^(data:image\/svg\+xml;utf8,)(.*)/;
function createTintColorSVG(tintColor, id) {
  return tintColor && id != null ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("svg", {
    style: {
      position: "absolute",
      height: 0,
      visibility: "hidden",
      width: 0
    },
    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("defs", {
      children: (
        /* @ts-ignore */
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("filter", {
          id: "tint-".concat(id),
          suppressHydrationWarning: !0,
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("feFlood", {
              floodColor: "".concat(tintColor)
            }, tintColor),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("feComposite", {
              in2: "SourceAlpha",
              operator: "atop"
            })
          ]
        })
      )
    })
  }) : null;
}
function getFlatStyle(style, blurRadius, filterId) {
  var flatStyle = import_react_native_web_internals.StyleSheet.flatten(style), filter = flatStyle.filter, resizeMode = flatStyle.resizeMode, shadowOffset = flatStyle.shadowOffset, tintColor = flatStyle.tintColor, filters = [], _filter = null;
  if (filter && filters.push(filter), blurRadius && filters.push("blur(".concat(blurRadius, "px)")), shadowOffset) {
    var shadowString = (0, import_react_native_web_internals.createBoxShadowValue)(flatStyle);
    shadowString && filters.push("drop-shadow(".concat(shadowString, ")"));
  }
  return tintColor && filterId != null && filters.push("url(#tint-".concat(filterId, ")")), filters.length > 0 && (_filter = filters.join(" ")), delete flatStyle.blurRadius, delete flatStyle.shadowColor, delete flatStyle.shadowOpacity, delete flatStyle.shadowOffset, delete flatStyle.shadowRadius, delete flatStyle.tintColor, delete flatStyle.overlayColor, delete flatStyle.resizeMode, [
    flatStyle,
    resizeMode,
    _filter,
    tintColor
  ];
}
function resolveAssetDimensions(source) {
  if (typeof source == "number") {
    var _getAssetByID = (0, import_react_native_web_internals2.getAssetByID)(source), height = _getAssetByID.height, width = _getAssetByID.width;
    return {
      height,
      width
    };
  } else if (source != null && !Array.isArray(source) && typeof source == "object") {
    var height1 = source.height, width1 = source.width;
    return {
      height: height1,
      width: width1
    };
  }
}
function resolveAssetUri(source) {
  var uri = null;
  if (typeof source == "number") {
    var asset = (0, import_react_native_web_internals2.getAssetByID)(source), scale = asset.scales[0];
    if (asset.scales.length > 1) {
      var preferredScale = import_PixelRatio.default.get();
      scale = asset.scales.reduce(function(prev, curr) {
        return Math.abs(curr - preferredScale) < Math.abs(prev - preferredScale) ? curr : prev;
      });
    }
    var scaleSuffix = scale !== 1 ? "@".concat(scale, "x") : "";
    uri = asset ? "".concat(asset.httpServerLocation, "/").concat(asset.name).concat(scaleSuffix, ".").concat(asset.type) : "";
  } else typeof source == "string" ? uri = source : source && typeof source.uri == "string" && (uri = source.uri);
  if (uri) {
    var match = uri.match(svgDataUriPattern);
    if (match) {
      var _match = _sliced_to_array(match, 3), prefix = _match[1], svg = _match[2], encodedSvg = encodeURIComponent(svg);
      return "".concat(prefix).concat(encodedSvg);
    }
  }
  return uri;
}
var Image = /* @__PURE__ */ React.forwardRef(function(props, ref) {
  var getBackgroundSize = function() {
    if (hiddenImageRef.current != null && (resizeMode === "center" || resizeMode === "repeat")) {
      var _hiddenImageRef_current = hiddenImageRef.current, naturalHeight = _hiddenImageRef_current.naturalHeight, naturalWidth = _hiddenImageRef_current.naturalWidth, height = layout.height, width = layout.width;
      if (naturalHeight && naturalWidth && height && width) {
        var scaleFactor = Math.min(1, width / naturalWidth, height / naturalHeight), x = Math.ceil(scaleFactor * naturalWidth), y = Math.ceil(scaleFactor * naturalHeight);
        return "".concat(x, "px ").concat(y, "px");
      }
    }
  }, handleLayout = function(e) {
    if (resizeMode === "center" || resizeMode === "repeat" || onLayout) {
      var layout2 = e.nativeEvent.layout;
      onLayout && onLayout(e), updateLayout(layout2);
    }
  }, accessibilityLabel = props.accessibilityLabel, blurRadius = props.blurRadius, defaultSource = props.defaultSource, draggable = props.draggable, onError = props.onError, onLayout = props.onLayout, onLoad = props.onLoad, onLoadEnd = props.onLoadEnd, onLoadStart = props.onLoadStart, pointerEvents = props.pointerEvents, source = props.source, style = props.style, rest = _object_without_properties(props, [
    "accessibilityLabel",
    "blurRadius",
    "defaultSource",
    "draggable",
    "onError",
    "onLayout",
    "onLoad",
    "onLoadEnd",
    "onLoadStart",
    "pointerEvents",
    "source",
    "style"
  ]);
  if (process.env.NODE_ENV !== "production" && props.children)
    throw new Error("The <Image> component cannot contain children. If you want to render content on top of the image, consider using the <ImageBackground> component or absolute positioning.");
  var _React_useState = _sliced_to_array(React.useState(function() {
    var uri2 = resolveAssetUri(source);
    if (uri2 != null) {
      var isLoaded = import_react_native_web_internals2.ImageLoader.has(uri2);
      if (isLoaded)
        return LOADED;
    }
    return IDLE;
  }), 2), state = _React_useState[0], updateState = _React_useState[1], _React_useState1 = _sliced_to_array(React.useState({}), 2), layout = _React_useState1[0], updateLayout = _React_useState1[1], hasTextAncestor = React.useContext(import_react_native_web_internals.TextAncestorContext), hiddenImageRef = React.useRef(null), filterRef = React.useRef(_filterId++), requestRef = React.useRef(null), shouldDisplaySource = state === LOADED || state === LOADING && defaultSource == null, _getFlatStyle = _sliced_to_array(getFlatStyle({}, blurRadius, filterRef.current), 4), flatStyle = _getFlatStyle[0], _resizeMode = _getFlatStyle[1], filter = _getFlatStyle[2], tintColor = _getFlatStyle[3], resizeMode = props.resizeMode || _resizeMode || "cover", selectedSource = shouldDisplaySource ? source : defaultSource, displayImageUri = resolveAssetUri(selectedSource), imageSizeStyle = resolveAssetDimensions(selectedSource), backgroundImage = displayImageUri ? 'url("'.concat(displayImageUri, '")') : null, backgroundSize = getBackgroundSize(), hiddenImage = displayImageUri ? (0, import_createElement.default)("img", {
    alt: accessibilityLabel || "",
    style: styles.accessibilityImage$raw,
    draggable: draggable || !1,
    ref: hiddenImageRef,
    src: displayImageUri
  }) : null, uri = resolveAssetUri(source);
  return React.useEffect(function() {
    var abortPendingRequest = function() {
      requestRef.current != null && (import_react_native_web_internals2.ImageLoader.abort(requestRef.current), requestRef.current = null);
    };
    return abortPendingRequest(), uri != null && (updateState(LOADING), onLoadStart && onLoadStart(), requestRef.current = import_react_native_web_internals2.ImageLoader.load(uri, function(e) {
      updateState(LOADED), onLoad && onLoad(e), onLoadEnd && onLoadEnd();
    }, function() {
      updateState(ERRORED), onError && onError({
        nativeEvent: {
          error: "Failed to load resource ".concat(uri, " (404)")
        }
      }), onLoadEnd && onLoadEnd();
    })), abortPendingRequest;
  }, [
    uri,
    requestRef,
    updateState,
    onError,
    onLoad,
    onLoadEnd,
    onLoadStart
  ]), /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_View.default, _object_spread_props(_object_spread({}, rest), {
    accessibilityLabel,
    onLayout: handleLayout,
    pointerEvents,
    ref,
    style: [
      style,
      styles.root,
      hasTextAncestor && styles.inline,
      imageSizeStyle,
      flatStyle
    ],
    children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_View.default, {
        style: _to_consumable_array([].concat(styles.image)).concat([
          resizeModeStyles[resizeMode],
          {
            backgroundImage,
            filter
          },
          backgroundSize != null && {
            backgroundSize
          }
        ]),
        // @ts-ignore
        suppressHydrationWarning: !0
      }),
      hiddenImage,
      createTintColorSVG(tintColor, filterRef.current)
    ]
  }));
});
Image.displayName = "Image";
var ImageWithStatics = Image;
ImageWithStatics.getSize = function(uri, success, failure) {
  import_react_native_web_internals2.ImageLoader.getSize(uri, success, failure);
};
ImageWithStatics.prefetch = function(uri) {
  return import_react_native_web_internals2.ImageLoader.prefetch(uri);
};
ImageWithStatics.queryCache = function(uris) {
  return import_react_native_web_internals2.ImageLoader.queryCache(uris);
};
var styles = import_react_native_web_internals.StyleSheet.create({
  root: {
    flexBasis: "auto",
    overflow: "hidden",
    zIndex: 0
  },
  inline: {
    display: "inline-flex"
  },
  image: _object_spread_props(_object_spread({}, import_react_native_web_internals.StyleSheet.absoluteFillObject), {
    backgroundColor: "transparent",
    backgroundPosition: "center",
    backgroundRepeat: "no-repeat",
    backgroundSize: "cover",
    height: "100%",
    width: "100%",
    zIndex: -1
  }),
  accessibilityImage$raw: _object_spread_props(_object_spread({}, import_react_native_web_internals.StyleSheet.absoluteFillObject), {
    height: "100%",
    opacity: 0,
    width: "100%",
    zIndex: -1
  })
}), resizeModeStyles = import_react_native_web_internals.StyleSheet.create({
  center: {
    backgroundSize: "auto"
  },
  contain: {
    backgroundSize: "contain"
  },
  cover: {
    backgroundSize: "cover"
  },
  none: {
    backgroundPosition: "0",
    backgroundSize: "auto"
  },
  repeat: {
    backgroundPosition: "0",
    backgroundRepeat: "repeat",
    backgroundSize: "auto"
  },
  stretch: {
    backgroundSize: "100% 100%"
  }
}), Image_default = ImageWithStatics;
