// @tamagui/focusable/dist/cjs/focusableInputHOC.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var focusableInputHOC_exports = {};
__export(focusableInputHOC_exports, {
  focusableInputHOC: () => focusableInputHOC,
  useFocusable: () => useFocusable
});
module.exports = __toCommonJS(focusableInputHOC_exports);
var import_jsx_runtime = require("react/jsx-runtime"), import_compose_refs = require("@tamagui/compose-refs"), import_web = require("@tamagui/web"), import_react = require("react"), import_registerFocusable = require("./registerFocusable");
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function useFocusable(param) {
  var isInput = param.isInput, props = param.props, ref = param.ref, id = props.id, onChangeText = props.onChangeText, value = props.value, defaultValue = props.defaultValue, inputValue = (0, import_react.useRef)(value || defaultValue || ""), unregisterFocusable = (0, import_react.useRef)(), inputRef = (0, import_react.useCallback)(function(input) {
    var _unregisterFocusable_current;
    id && input && ((_unregisterFocusable_current = unregisterFocusable.current) === null || _unregisterFocusable_current === void 0 || _unregisterFocusable_current.call(unregisterFocusable), unregisterFocusable.current = (0, import_registerFocusable.registerFocusable)(id, _object_spread({
      focus: input.focus
    }, isInput && {
      // react-native doesn't support programmatic .select()
      focusAndSelect: function() {
        input.focus(), input.setSelection && typeof inputValue.current == "string" && input.setSelection(0, inputValue.current.length);
      }
    })));
  }, [
    isInput,
    id
  ]), combinedRefs = (0, import_compose_refs.composeRefs)(ref, inputRef);
  return (0, import_react.useEffect)(function() {
    return function() {
      var _unregisterFocusable_current;
      (_unregisterFocusable_current = unregisterFocusable.current) === null || _unregisterFocusable_current === void 0 || _unregisterFocusable_current.call(unregisterFocusable);
    };
  }, []), {
    ref: combinedRefs,
    onChangeText: (0, import_web.useEvent)(function(value2) {
      inputValue.current = value2, onChangeText == null || onChangeText(value2);
    })
  };
}
function focusableInputHOC(Component) {
  return Component.styleable(function(props, ref) {
    var _Component_staticConfig, isInput = (_Component_staticConfig = Component.staticConfig) === null || _Component_staticConfig === void 0 ? void 0 : _Component_staticConfig.isInput, _useFocusable = useFocusable({
      ref,
      props,
      isInput
    }), combinedRef = _useFocusable.ref, onChangeText = _useFocusable.onChangeText, finalProps = isInput ? _object_spread_props(_object_spread({}, props), {
      onChangeText
    }) : props;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Component, _object_spread({
      ref: combinedRef
    }, finalProps));
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  focusableInputHOC,
  useFocusable
});
