// react-native-web-lite/dist/esm/Modal/ModalContent.native.js
import * as React from "react";
import { StyleSheet, canUseDOM } from "react-native-web-internals";
import View from "../View";
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function _object_without_properties(source, excluded) {
  if (source == null) return {};
  var target = _object_without_properties_loose(source, excluded), key, i;
  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
    for (i = 0; i < sourceSymbolKeys.length; i++)
      key = sourceSymbolKeys[i], !(excluded.indexOf(key) >= 0) && Object.prototype.propertyIsEnumerable.call(source, key) && (target[key] = source[key]);
  }
  return target;
}
function _object_without_properties_loose(source, excluded) {
  if (source == null) return {};
  var target = {}, sourceKeys = Object.keys(source), key, i;
  for (i = 0; i < sourceKeys.length; i++)
    key = sourceKeys[i], !(excluded.indexOf(key) >= 0) && (target[key] = source[key]);
  return target;
}
var ModalContent = /* @__PURE__ */ React.forwardRef(function(props, forwardedRef) {
  var active = props.active, children = props.children, onRequestClose = props.onRequestClose, transparent = props.transparent, rest = _object_without_properties(props, [
    "active",
    "children",
    "onRequestClose",
    "transparent"
  ]);
  React.useEffect(function() {
    if (canUseDOM) {
      var closeOnEscape = function(e) {
        active && e.key === "Escape" && (e.stopPropagation(), onRequestClose && onRequestClose());
      };
      return document.addEventListener("keyup", closeOnEscape, !1), function() {
        return document.removeEventListener("keyup", closeOnEscape, !1);
      };
    }
  }, [
    active,
    onRequestClose
  ]);
  var style = React.useMemo(function() {
    return [
      styles.modal,
      transparent ? styles.modalTransparent : styles.modalOpaque
    ];
  }, [
    transparent
  ]);
  return /* @__PURE__ */ React.createElement(View, _object_spread_props(_object_spread({}, rest), {
    accessibilityModal: !0,
    accessibilityRole: active ? "dialog" : null,
    ref: forwardedRef,
    style
  }), /* @__PURE__ */ React.createElement(View, {
    style: styles.container
  }, children));
}), styles = StyleSheet.create({
  modal: {
    position: "fixed",
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  },
  modalTransparent: {
    backgroundColor: "transparent"
  },
  modalOpaque: {
    backgroundColor: "white"
  },
  container: {
    top: 0,
    flex: 1
  }
}), ModalContent_default = ModalContent;
export {
  ModalContent_default as default
};
