// @tamagui/sheet/dist/cjs/createSheet.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var createSheet_exports = {};
__export(createSheet_exports, {
  createSheet: () => createSheet
});
module.exports = __toCommonJS(createSheet_exports);
var import_jsx_runtime = require("react/jsx-runtime"), import_compose_refs = require("@tamagui/compose-refs"), import_constants = require("@tamagui/constants"), import_core = require("@tamagui/core"), import_helpers = require("@tamagui/helpers"), import_remove_scroll = require("@tamagui/remove-scroll"), import_use_did_finish_ssr = require("@tamagui/use-did-finish-ssr"), import_react = require("react"), import_react_native = require("react-native"), import_constants2 = require("./constants"), import_nativeSheet = require("./nativeSheet"), import_SheetContext = require("./SheetContext"), import_SheetImplementationCustom = require("./SheetImplementationCustom"), import_SheetScrollView = require("./SheetScrollView"), import_useSheetController = require("./useSheetController"), import_useSheetOffscreenSize = require("./useSheetOffscreenSize");
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function _object_without_properties(source, excluded) {
  if (source == null) return {};
  var target = _object_without_properties_loose(source, excluded), key, i;
  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
    for (i = 0; i < sourceSymbolKeys.length; i++)
      key = sourceSymbolKeys[i], !(excluded.indexOf(key) >= 0) && Object.prototype.propertyIsEnumerable.call(source, key) && (target[key] = source[key]);
  }
  return target;
}
function _object_without_properties_loose(source, excluded) {
  if (source == null) return {};
  var target = {}, sourceKeys = Object.keys(source), key, i;
  for (i = 0; i < sourceKeys.length; i++)
    key = sourceKeys[i], !(excluded.indexOf(key) >= 0) && (target[key] = source[key]);
  return target;
}
function createSheet(param) {
  var Handle = param.Handle, Frame = param.Frame, Overlay = param.Overlay, SheetHandle = Handle.styleable(function(_param, forwardedRef) {
    var __scopeSheet = _param.__scopeSheet, props = _object_without_properties(_param, [
      "__scopeSheet"
    ]), context = (0, import_SheetContext.useSheetContext)(import_constants2.SHEET_HANDLE_NAME, __scopeSheet), composedRef = (0, import_compose_refs.useComposedRefs)(context.handleRef, forwardedRef);
    return context.onlyShowFrame ? null : (
      // @ts-ignore
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Handle, _object_spread({
        ref: composedRef,
        onPress: function() {
          var max = context.snapPoints.length + (context.dismissOnSnapToBottom ? -1 : 0), nextPos = (context.position + 1) % max;
          context.setPosition(nextPos);
        },
        open: context.open
      }, props))
    );
  }), SheetOverlay = Overlay.extractable(/* @__PURE__ */ (0, import_react.memo)(function(propsIn) {
    var __scopeSheet = propsIn.__scopeSheet, props = _object_without_properties(propsIn, [
      "__scopeSheet"
    ]), context = (0, import_SheetContext.useSheetContext)(import_constants2.SHEET_OVERLAY_NAME, __scopeSheet), element = (0, import_react.useMemo)(function() {
      return (
        // @ts-ignore
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Overlay, _object_spread_props(_object_spread({}, props), {
          onPress: (0, import_helpers.composeEventHandlers)(props.onPress, context.dismissOnOverlayPress ? function() {
            context.setOpen(!1);
          } : void 0)
        }))
      );
    }, [
      props.onPress,
      context.dismissOnOverlayPress
    ]);
    return (0, import_constants.useIsomorphicLayoutEffect)(function() {
      var _context_onOverlayComponent;
      (_context_onOverlayComponent = context.onOverlayComponent) === null || _context_onOverlayComponent === void 0 || _context_onOverlayComponent.call(context, element);
    }, [
      element
    ]), context.onlyShowFrame, null;
  })), SheetFrame = Frame.extractable(/* @__PURE__ */ (0, import_react.forwardRef)(function(_param, forwardedRef) {
    var __scopeSheet = _param.__scopeSheet, adjustPaddingForOffscreenContent = _param.adjustPaddingForOffscreenContent, disableHideBottomOverflow = _param.disableHideBottomOverflow, children = _param.children, props = _object_without_properties(_param, [
      "__scopeSheet",
      "adjustPaddingForOffscreenContent",
      "disableHideBottomOverflow",
      "children"
    ]), context = (0, import_SheetContext.useSheetContext)(import_constants2.SHEET_NAME, __scopeSheet), hasFit = context.hasFit, removeScrollEnabled = context.removeScrollEnabled, frameSize = context.frameSize, contentRef = context.contentRef, open = context.open, composedContentRef = (0, import_compose_refs.useComposedRefs)(forwardedRef, contentRef), offscreenSize = (0, import_useSheetOffscreenSize.useSheetOffscreenSize)(context), sheetContents = (0, import_react.useMemo)(function() {
      return (
        // @ts-ignore
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(Frame, _object_spread_props(_object_spread({
          ref: composedContentRef,
          flex: hasFit ? 0 : 1,
          height: hasFit ? void 0 : frameSize,
          pointerEvents: open ? "auto" : "none"
        }, props), {
          children: [
            children,
            adjustPaddingForOffscreenContent && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_core.Stack, {
              "data-sheet-offscreen-pad": !0,
              height: offscreenSize,
              width: "100%"
            })
          ]
        }))
      );
    }, [
      open,
      props,
      frameSize,
      offscreenSize,
      adjustPaddingForOffscreenContent,
      hasFit
    ]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_remove_scroll.RemoveScroll, {
          forwardProps: !0,
          enabled: removeScrollEnabled,
          allowPinchZoom: !0,
          shards: [
            contentRef
          ],
          // causes lots of bugs on touch web on site
          removeScrollBar: !1,
          children: sheetContents
        }),
        /* below frame hide when bouncing past 100% */
        !disableHideBottomOverflow && // @ts-ignore
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Frame, _object_spread_props(_object_spread({}, props), {
          componentName: "SheetCover",
          // biome-ignore lint/correctness/noChildrenProp: <explanation>
          children: null,
          position: "absolute",
          bottom: "-100%",
          zIndex: -1,
          height: context.frameSize,
          left: 0,
          right: 0,
          borderWidth: 0,
          borderRadius: 0,
          shadowOpacity: 0
        }))
      ]
    });
  })), Sheet = /* @__PURE__ */ (0, import_react.forwardRef)(function(props, ref) {
    var hydrated = (0, import_use_did_finish_ssr.useDidFinishSSR)(), isShowingNonSheet = (0, import_useSheetController.useSheetController)().isShowingNonSheet, SheetImplementation = import_SheetImplementationCustom.SheetImplementationCustom;
    if (props.native && import_react_native.Platform.OS === "ios") {
      var impl = (0, import_nativeSheet.getNativeSheet)("ios");
      impl && (SheetImplementation = impl);
    }
    return isShowingNonSheet || !hydrated ? null : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SheetImplementation, _object_spread({
      ref
    }, props));
  }), components = {
    Frame: SheetFrame,
    Overlay: SheetOverlay,
    Handle: SheetHandle,
    ScrollView: import_SheetScrollView.SheetScrollView
  }, Controlled = (0, import_helpers.withStaticProperties)(Sheet, components);
  return (0, import_helpers.withStaticProperties)(Sheet, _object_spread_props(_object_spread({}, components), {
    Controlled
  }));
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createSheet
});
