// matrix-js-sdk/lib/http-api/errors.js
/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Construct a generic HTTP error. This is a JavaScript Error with additional information
 * specific to HTTP responses.
 * @param msg - The error message to include.
 * @param httpStatus - The HTTP response status code.
 */
export class HTTPError extends Error {
  constructor(msg, httpStatus) {
    super(msg);
    this.httpStatus = httpStatus;
  }
}
export class MatrixError extends HTTPError {
  // The Matrix 'errcode' value, e.g. "M_FORBIDDEN".

  // The raw Matrix error JSON used to construct this object.

  /**
   * Construct a Matrix error. This is a JavaScript Error with additional
   * information specific to the standard Matrix error response.
   * @param errorJson - The Matrix error JSON returned from the homeserver.
   * @param httpStatus - The numeric HTTP status code given
   */
  constructor() {
    var errorJson = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var httpStatus = arguments.length > 1 ? arguments[1] : undefined;
    var url = arguments.length > 2 ? arguments[2] : undefined;
    var event = arguments.length > 3 ? arguments[3] : undefined;
    var message = errorJson.error || "Unknown message";
    if (httpStatus) {
      message = "[".concat(httpStatus, "] ").concat(message);
    }
    if (url) {
      message = "".concat(message, " (").concat(url, ")");
    }
    super("MatrixError: ".concat(message), httpStatus);
    this.httpStatus = httpStatus;
    this.url = url;
    this.event = event;
    this.errcode = errorJson.errcode;
    this.name = errorJson.errcode || "Unknown error code";
    this.data = errorJson;
  }
}

/**
 * Construct a ConnectionError. This is a JavaScript Error indicating
 * that a request failed because of some error with the connection, either
 * CORS was not correctly configured on the server, the server didn't response,
 * the request timed out, or the internet connection on the client side went down.
 */
export class ConnectionError extends Error {
  constructor(message, cause) {
    super(message + (cause ? ": ".concat(cause.message) : ""));
  }
  get name() {
    return "ConnectionError";
  }
}
