// @tamagui/radio-group/dist/cjs/createRadioGroup.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var createRadioGroup_exports = {};
__export(createRadioGroup_exports, {
  createRadioGroup: () => createRadioGroup
});
module.exports = __toCommonJS(createRadioGroup_exports);
var import_jsx_runtime = require("react/jsx-runtime"), import_core = require("@tamagui/core"), import_RadioGroup = require("./RadioGroup"), import_radio_headless = require("@tamagui/radio-headless"), import_roving_focus = require("@tamagui/roving-focus"), import_react = require("react");
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function _object_without_properties(source, excluded) {
  if (source == null) return {};
  var target = _object_without_properties_loose(source, excluded), key, i;
  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
    for (i = 0; i < sourceSymbolKeys.length; i++)
      key = sourceSymbolKeys[i], !(excluded.indexOf(key) >= 0) && Object.prototype.propertyIsEnumerable.call(source, key) && (target[key] = source[key]);
  }
  return target;
}
function _object_without_properties_loose(source, excluded) {
  if (source == null) return {};
  var target = {}, sourceKeys = Object.keys(source), key, i;
  for (i = 0; i < sourceKeys.length; i++)
    key = sourceKeys[i], !(excluded.indexOf(key) >= 0) && (target[key] = source[key]);
  return target;
}
var ensureContext = function(x) {
  x.context || (x.context = RadioGroupContext);
}, RadioGroupContext = /* @__PURE__ */ (0, import_react.createContext)({}), RadioGroupItemContext = /* @__PURE__ */ (0, import_react.createContext)({
  checked: !1,
  disabled: !1
});
function createRadioGroup(createProps) {
  var disableActiveTheme = createProps.disableActiveTheme, _createProps_Frame = createProps.Frame, Frame = _createProps_Frame === void 0 ? import_RadioGroup.RadioGroupFrame : _createProps_Frame, _createProps_Indicator = createProps.Indicator, Indicator = _createProps_Indicator === void 0 ? import_RadioGroup.RadioGroupIndicatorFrame : _createProps_Indicator, _createProps_Item = createProps.Item, Item = _createProps_Item === void 0 ? import_RadioGroup.RadioGroupItemFrame : _createProps_Item;
  ensureContext(Frame), ensureContext(Indicator), ensureContext(Item);
  var RadioGroupImp = Frame.styleable(function(props, ref) {
    var value = props.value, defaultValue = props.defaultValue, onValueChange = props.onValueChange, _props_required = props.required, required = _props_required === void 0 ? !1 : _props_required, _props_disabled = props.disabled, disabled = _props_disabled === void 0 ? !1 : _props_disabled, name = props.name, native = props.native, accentColor = props.accentColor, _props_orientation = props.orientation, orientation = _props_orientation === void 0 ? "vertical" : _props_orientation, rest = _object_without_properties(props, [
      "value",
      "defaultValue",
      "onValueChange",
      "required",
      "disabled",
      "name",
      "native",
      "accentColor",
      "orientation"
    ]), _useRadioGroup = (0, import_radio_headless.useRadioGroup)({
      orientation,
      name,
      defaultValue,
      value,
      onValueChange,
      required,
      disabled,
      native,
      accentColor
    }), providerValue = _useRadioGroup.providerValue, frameAttrs = _useRadioGroup.frameAttrs, rovingFocusGroupAttrs = _useRadioGroup.rovingFocusGroupAttrs;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(RadioGroupContext.Provider, {
      value: providerValue,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_roving_focus.RovingFocusGroup, _object_spread_props(_object_spread({}, rovingFocusGroupAttrs), {
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RadioGroup.RadioGroupFrame, _object_spread(_object_spread_props(_object_spread({}, frameAttrs), {
          ref
        }), rest))
      }))
    });
  }), RadioGroupItemImp = Item.styleable(function(props, ref) {
    var value = props.value, labelledBy = props.labelledBy, onPress = props.onPress, onKeyDown = props.onKeyDown, disabled = props.disabled, id = props.id, rest = _object_without_properties(props, [
      "value",
      "labelledBy",
      "onPress",
      "onKeyDown",
      "disabled",
      "id"
    ]), _useRadioGroupItem = (0, import_radio_headless.useRadioGroupItem)({
      radioGroupContext: RadioGroupContext,
      value,
      id,
      labelledBy,
      disabled,
      onPress,
      onKeyDown
    }), providerValue = _useRadioGroupItem.providerValue, bubbleInput = _useRadioGroupItem.bubbleInput, rovingFocusGroupAttrs = _useRadioGroupItem.rovingFocusGroupAttrs, frameAttrs = _useRadioGroupItem.frameAttrs, isFormControl = _useRadioGroupItem.isFormControl, native = _useRadioGroupItem.native;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(RadioGroupItemContext.Provider, {
      value: providerValue,
      children: import_core.isWeb && native ? bubbleInput : /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_roving_focus.RovingFocusGroup.Item, _object_spread_props(_object_spread({}, rovingFocusGroupAttrs), {
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RadioGroup.RadioGroupItemFrame, _object_spread(_object_spread_props(_object_spread({}, frameAttrs), {
              ref
            }), rest))
          })),
          isFormControl && bubbleInput
        ]
      })
    });
  });
  RadioGroupItemImp.displayName = "RadioGroupItem";
  var RadioIndicator = Indicator.styleable(function(props, forwardedRef) {
    var forceMount = props.forceMount, disabled = props.disabled, indicatorProps = _object_without_properties(props, [
      "forceMount",
      "disabled"
    ]), _useRadioGroupItemIndicator = (0, import_radio_headless.useRadioGroupItemIndicator)({
      radioGroupItemContext: RadioGroupItemContext,
      disabled
    }), checked = _useRadioGroupItemIndicator.checked, useIndicatorRest = _object_without_properties(_useRadioGroupItemIndicator, [
      "checked"
    ]);
    return forceMount || checked ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Indicator, _object_spread(_object_spread_props(_object_spread({}, useIndicatorRest), {
      ref: forwardedRef
    }), indicatorProps)) : null;
  });
  RadioIndicator.displayName = "RadioIndicator";
  var RadioGroup = (0, import_core.withStaticProperties)(RadioGroupImp, {
    Item: RadioGroupItemImp,
    Indicator: RadioIndicator
  });
  return RadioGroup.displayName = "RadioGroup", RadioGroup;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createRadioGroup
});
