// react-native-web-lite/dist/esm/vendor/react-native/Animated/useAnimatedProps.native.js
import _createForOfIteratorHelperLoose from "@babel/runtime/helpers/createForOfIteratorHelperLoose";
import { useLayoutEffect as _, startTransition, useCallback, useEffect, useMemo, useReducer, useRef } from "react";
import useRefEffect from "../useRefEffect";
import { AnimatedEvent } from "./AnimatedEvent";
import NativeAnimatedHelper from "./NativeAnimatedHelper";
import AnimatedProps from "./nodes/AnimatedProps";
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _instanceof(left, right) {
  return right != null && typeof Symbol < "u" && right[Symbol.hasInstance] ? !!right[Symbol.hasInstance](left) : left instanceof right;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
var useLayoutEffect = typeof window > "u" ? useEffect : _;
function useAnimatedProps(props) {
  var _useReducer = useReducer(function(count) {
    return count + 1;
  }, 0), scheduleUpdate = _useReducer[1], onUpdateRef = useRef(null), node = useMemo(function() {
    return new AnimatedProps(props, function() {
      return onUpdateRef.current == null ? void 0 : onUpdateRef.current();
    });
  }, [
    props
  ]);
  useAnimatedPropsLifecycle(node);
  var refEffect = useCallback(function(instance) {
    node.setNativeView(instance), onUpdateRef.current = function() {
      startTransition(function() {
        scheduleUpdate();
      });
    };
    var target = getEventTarget(instance), events = [];
    for (var propName in props) {
      var propValue = props[propName];
      _instanceof(propValue, AnimatedEvent) && propValue.__isNative && (propValue.__attach(target, propName), events.push([
        propName,
        propValue
      ]));
    }
    return function() {
      onUpdateRef.current = null;
      for (var _iterator = _createForOfIteratorHelperLoose(events), _step; !(_step = _iterator()).done; ) {
        var _step$value = _step.value, _propName = _step$value[0], _propValue = _step$value[1];
        _propValue.__detach(target, _propName);
      }
    };
  }, [
    props,
    node
  ]), callbackRef = useRefEffect(refEffect);
  return [
    reduceAnimatedProps(node),
    callbackRef
  ];
}
function reduceAnimatedProps(node) {
  return _object_spread_props(_object_spread({}, node.__getValue()), {
    collapsable: !1
  });
}
function useAnimatedPropsLifecycle(node) {
  var prevNodeRef = useRef(null), isUnmountingRef = useRef(!1);
  useEffect(function() {
    NativeAnimatedHelper.API.flushQueue();
  }), useLayoutEffect(function() {
    return isUnmountingRef.current = !1, function() {
      isUnmountingRef.current = !0;
    };
  }, []), useLayoutEffect(function() {
    if (node.__attach(), prevNodeRef.current != null) {
      var prevNode = prevNodeRef.current;
      prevNode.__restoreDefaultValues(), prevNode.__detach(), prevNodeRef.current = null;
    }
    return function() {
      isUnmountingRef.current ? node.__detach() : prevNodeRef.current = node;
    };
  }, [
    node
  ]);
}
function getEventTarget(instance) {
  return typeof instance == "object" && typeof (instance == null ? void 0 : instance.getScrollableNode) == "function" ? (
    // $FlowFixMe[incompatible-use] - Legacy instance assumptions.
    instance.getScrollableNode()
  ) : instance;
}
export {
  useAnimatedProps as default
};
